### PR TITLE
Implement WebView evalJS and local scripting

### PR DESCRIPTION
**Pull request by infosia**
[TIMOB-20289](https://jira.appcelerator.org/browse/TIMOB-20289)

Implementation for Ti.UI.WebView Titanium API in the HTML as known as "local script" as described in the [API doc](http://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.WebView). Especially `Ti.App.addEventListener`, `Ti.App.fireEvent` and Ti.API logging functions.

See also: [Communication Between WebViews and Titanium](https://wiki.appcelerator.org/display/guides2/Communication+Between+WebViews+and+Titanium)

- app.js
```javascript
var win = Ti.UI.createWindow();
var view = Ti.UI.createView({backgroundColor:'blue'});
var webview = Ti.UI.createWebView({
    url: 'WebViewTestPage.htm',
    height: '80%',
    width: Ti.UI.FILL,
    bottom: 0
});
var button = Ti.UI.createButton({
    title: 'fromTitanium',
    height: '10%',
    width: Ti.UI.FILL,
    top: 0, left: 0
});
button.addEventListener('click', function (e) {
    Ti.App.fireEvent('app:fromTitanium', { message: 'event fired from Titanium, handled in WebView' });
});
Ti.App.addEventListener('app:fromWebView', function (e) {
    alert(e.message);
});

view.add(webview);
view.add(button);
win.add(view);
win.open();
```

* WebViewTestPage.htm
```html
<html>
<head>
    <script>
        Ti.App.addEventListener("app:fromTitanium", function(e) {
            Ti.API.info(e.message);
        });
    </script>
</head>
<body>
    <button onclick="Ti.App.fireEvent('app:fromWebView', { message: 'event fired from WebView, handled in Titanium' });">fromWebView</button>
</body>
</html>
```

This PR also includes asynchronous implementation for `Ti.WebView.evalJS(code, callback)` for Windows. Async callback is needed because JavaScript evaluation on Windows WebView needs to be executed with async.

```javascript
var w = Ti.UI.createWindow({
    backgroundColor: 'blue',
    html: '<html><body>Hello, World!</body></html>'
});
var webview = Ti.UI.createWebView();
webview.addEventListener('load', function () {
    Ti.API.info('WebView load');
    webview.evalJS('"WebView.evalJS.TEST";', function (result) {
        Ti.API.info('Should be true: ' + ('WebView.evalJS.TEST' == result));
    });
});

w.add(webview);
w.open();
```

### CODE REVIEW

File: Source/UI/include/TitaniumWindows/UI/WebView.hpp [hunk 1]
```diff
@@ -74,6 +77,8 @@ namespace TitaniumWindows
 			Windows::Foundation::EventRegistrationToken load_event__;
 			Windows::Foundation::EventRegistrationToken beforeload_event__;
 			Windows::Foundation::EventRegistrationToken script_event__;
+
+			std::unordered_set<std::string> app_event_names;
```
Comment (sgtcoolguy) [inline Source/UI/include/TitaniumWindows/UI/WebView.hpp]: Build is throwing a MSBuild warning her:
```cpp
#pragma warning(push)
#pragma warning(disable : 4251)
			std::unordered_set<std::string> app_event_names;
#pragma warning(pop)
```

File: Source/TitaniumKit/src/UI/WebView.cpp [hunk 2]
```diff
@@ -284,6 +299,41 @@ namespace Titanium
...
+		//
+		TITANIUM_FUNCTION(WebView, _executeListener)
+		{
+			if (arguments.size() < 2) {
+				TITANIUM_LOG_WARN("WebView._executeListener: Invalid arguments");
+				return get_context().CreateUndefined();
+			}
+
+			auto callee = arguments.at(1);
+			auto tostring = static_cast<std::string>(callee);
```
Comment (sgtcoolguy) [inline Source/TitaniumKit/src/UI/WebView.cpp]: Why convert to a string here?

Use our shorthand macro for converting to object?
```cpp
ENSURE_OBJECT_AT_INDEX(callee, 1);
```

File: Source/TitaniumKit/src/UI/WebView.cpp [hunk 3]
```diff
@@ -284,6 +299,41 @@ namespace Titanium
...
+			TITANIUM_ASSERT(callee.IsObject());
+			auto webview = static_cast<JSObject>(callee).GetProperty("webview");
+			if (!webview.IsObject()) {
+				TITANIUM_LOG_WARN("WebView._executeListener: Can't find this object");
+				return get_context().CreateUndefined();
+			}
+			auto webview_ptr = static_cast<JSObject>(webview).GetPrivate<Titanium::UI::WebView>();
+			TITANIUM_ASSERT(webview_ptr);
+
+			auto data = arguments.at(0);
```
Comment (sgtcoolguy) [inline Source/TitaniumKit/src/UI/WebView.cpp]: Use our shorthand macro for converting to object?
```cpp
ENSURE_OBJECT_AT_INDEX(data, 0);
```

File: Source/UI/include/TitaniumWindows/UI/WebView.hpp [hunk 4]
```diff
@@ -44,16 +44,19 @@ namespace TitaniumWindows
...
 			virtual void enableEvent(const std::string& event_name)  TITANIUM_NOEXCEPT;
 			virtual	void disableEvent(const std::string& event_name) TITANIUM_NOEXCEPT;
 
-		private:
-
 			Titanium::UI::URL_ERROR getUrlError(const Windows::Web::WebErrorStatus status) const TITANIUM_NOEXCEPT;
 			void initWebViewListeners() TITANIUM_NOEXCEPT;
+			void navigateWithLocalScript(const std::string& content);
 
+			virtual void set_html(const std::string& html) TITANIUM_NOEXCEPT override;
```
Comment (sgtcoolguy) [inline Source/UI/include/TitaniumWindows/UI/WebView.hpp]: Do we really need to override set_html (property setter) and setHtml (method)? Shouldn't the property better just call to the method with some default options (or no options) at the TitaniumKit level?

File: Source/UI/src/WebView.cpp [hunk 5]
```diff
@@ -73,13 +87,60 @@ namespace TitaniumWindows
 			return Titanium::UI::URL_ERROR::UNKNOWN;
 		}
 
+		std::string WebView::evalJS(const std::string& code, JSObject& callback) TITANIUM_NOEXCEPT
+		{
+			auto args = ref new Platform::Collections::Vector<Platform::String^>();
+			args->Append(TitaniumWindows::Utility::ConvertUTF8String(code));
+
+			// Always return empty string because WebView.InvokeScriptAsync only works with async on Windows
```
Comment (sgtcoolguy) [inline Source/UI/src/WebView.cpp]: move this comment down below just above return statement?

**Comment by infosia** [comment]
build win 10

**Pull request merged**